src/validation.cpp

```diff
@@ -4671,32 +4671,44 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                     continue;
             } catch (const std::exception&) {
                 // no valid block header found; don't complain
+                // (this happens at the end of every blk.dat file)
                 break;
             }
             try {
-                // read block
+                // read block header
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
-                blkdat.SetLimit(nBlockPos + nSize);
-                blkdat.SetPos(nBlockPos);
-                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                CBlock& block = *pblock;
-                blkdat >> block;
-                nRewind = blkdat.GetPos();
-
-                uint256 hash = block.GetHash();
+                CBlockHeader header;
+                blkdat >> header;
+
+                // Read the rest of the block (from the end of its header to the end
+                // of the block) to ensure it's in memory while holding cs_main (below).
+                blkdat.Skip(nBlockPos + nSize - blkdat.GetPos());
```

Comment:
@narula, this addresses your concern about holding `cs_main` while reading disk.